### PR TITLE
Enable quirky ANSI keys on Windows.

### LINKED ISSUE #1958: Missing Ctrl combo mappings on Windows
The keys Ctrl+2 (`^@`), Ctrl+Space (`^@`), Ctrl+/ (`^_`), Shift+Tab (`^[[Z`) are not implemented on Windows. These impact Neovim's usability in the Windows console - see https://github.com/neovim/neovim/issues/8435.

* **Version**: Not sure how to find this from neovim, but I can see they are still missing in master - see `src/win/tty.c: get_vt100_fn_key`.
* **Platform**: Windows 10 Insider 17746.

### PR DESCRIPTION
**Pull request by parkovski**
Fixes #1958. I've checked these bindings in plain conhost, conhost/tmux, and Windows Terminal/tmux.

### CODE REVIEW

**Comment by musm** [comment]
Great!

**Comment by parkovski** [comment]
Could help for someone not on a US keyboard layout to try this in their normal workflow. I tried to cover those cases but I use a US keyboard layout so I can't say for sure if there's not some quirks I missed.

Other than that, I've been using this patch daily in neovim for months with no issues. I hope this is a small enough/well documented enough change, because doing automated tests for all the variations of regular console/conpty/conpty over WSL interop/keyboard layouts is a pretty daunting project that I probably couldn't take on for a while.

**Comment by bnoordhuis** [comment]
@libuv/windows This PR is in need of review. It looks okay to me but I'm not qualified to judge.

**Comment by justinmk** [comment]
This is not stale, it is waiting on maintainer response.

File: src/win/tty.c [hunk 1]
```diff
@@ -852,6 +925,17 @@ void uv_process_tty_read_raw_req(uv_loop_t* loop, uv_tty_t* handle,
         const char* vt100;
         size_t prefix_len, vt100_len;
 
+        /* The key codes for the '/' and '_' keys depend on the user's keyboard
+         * layout.  Since the layout may change, and is actually very likely to
+         * change for devs that don't use EN-US by default, just look it up each
+         * loop and replace them with the ASCII codes, which get_vt100_fn_key
+         * knows how to translate. */
+        if (KEV.wVirtualKeyCode == LOBYTE(VkKeyScanA('/'))) {
```
Comment (vtjnash) [inline src/win/tty.c]: I'm not familiar with this system (or non EN-US keyboards). But why isn't it reporting `VK_DIVIDE`?

File: src/win/tty.c [hunk 2]
```diff
@@ -803,6 +836,46 @@ void uv_process_tty_read_raw_req(uv_loop_t* loop, uv_tty_t* handle,
...
 
+        /* Handle a couple of special cases here that may not hit the standard
+         * get_vt100_fn_key path. These match the behavior of zsh under
+         * WSL/conhost. There is a little bit of duplicate handling here and
+         * in get_vt100_fn_key. This is needed because of slight differences
+         * in the way conhost generates key presses directly or over a PTY. */
+        if (KEV.uChar.UnicodeChar == '\t') {
+          if (KEV.dwControlKeyState &
+              (LEFT_CTRL_PRESSED | RIGHT_CTRL_PRESSED)) {
+            continue;
```
Comment (vtjnash) [inline src/win/tty.c]:
```suggestion
            continue; /* Ignore C-tab */
```

File: src/win/tty.c [hunk 3]
```diff
@@ -803,6 +836,46 @@ void uv_process_tty_read_raw_req(uv_loop_t* loop, uv_tty_t* handle,
...
+            handle->tty.rd.last_key_offset = 0;
+            continue;
+          }
+        } else if (KEV.uChar.UnicodeChar == 0x1F) {
+          /* The Ctrl-_ key is a weird one. In conpty mode, we will see a key
+           * with UnicodeChar == 0x1F, which is handled here. With a direct
+           * console connection however, the UnicodeChar value is 0. The problem
+           * then becomes that the VK code of the '/' key depends on the user's
+           * keyboard layout. */
+          handle->tty.rd.last_key[prefix_len] = 0x1F;
```
Comment (vtjnash) [inline src/win/tty.c]: This seems the same as the default fall-through, but I don't know this aspect of the code very well. What am I missing?

File: src/win/tty.c (hunk 1 above)
Comment (parkovski) [reply src/win/tty.c]: On my keyboard, `VK_DIVIDE` is the numpad `/` key, but the one next to shift shows up as `VK_OEM_2` which MSDN says can represent different keys depending on layout. I forgot to handle `VK_DIVIDE` here, at least it's not working on my copy of Neovim with these changes, so I'll need to fix that.

Also see [here](https://github.com/microsoft/terminal/blob/322989d01701cc688593f4b3f06e535b77a8810f/src/terminal/input/terminalInput.cpp#L329) where they do the same in the official Windows terminal.

Comment (parkovski) [reply src/win/tty.c]: Actually it looks like I need to dig into this a little more, it's resolving to the wrong key on the Spanish keyboard layout.

File: src/win/tty.c (hunk 3 above)
Comment (parkovski) [reply src/win/tty.c]: This is indeed unnecessary, I'll take it out.

**Comment by parkovski** [comment]
So this is unfortunately more complicated than I'd thought (of course it is). At least for console apps, `VkScanKey` doesn't use the current keyboard layout, it uses whichever the app was started with, so you need to listen for `WM_INPUTLANGCHANGE` which requires a WndProc somewhere which libuv currently doesn't appear to have...

It seems to me like there are a couple of options here:

1. Load the current keyboard layout on every key press. This option doesn't feel that great to me, but...?
2. Do it the "right" way, creating a WndProc and caching the keyboard layout and/or the VK code for the `/` key when the language changes.
3. Pull the VK code check into a global that's checked at startup, commit this with the caveat that it doesn't handle keyboard changes correctly (really only affects the `/` key) and create a new issue for this.

File: src/win/tty.c (hunk 2 above)
Comment (parkovski) [reply src/win/tty.c]: Added this comment to my changes.

**Comment by justinmk** [comment]
> Pull the VK code check into a global that's checked at startup, commit this with the caveat that it doesn't handle keyboard changes correctly (really only affects the / key) and create a new issue for 

That one.

**Comment by vtjnash** [comment]
1 or 3 sounds good to me. We already can’t synchronize the stream exactly. I’d slightly lean towards doing 1 though and/or rereading it occasionally instead of at startup (eg whenever setting the mode to raw)

**Comment by musm** [comment]
plz don't

File: src/win/tty.c (hunk 3 above)
Comment (justinmk) [reply src/win/tty.c:873]: @parkovski did you get a chance to update this?

**Comment by vtjnash** [comment]
@parkovski were you still working on making this functional (https://github.com/libuv/libuv/pull/2535#discussion_r371508163)?

**Comment by musm** [comment]
bad bot

**Comment by parkovski** [comment]
Hey, life has been hectic for a while. I do want to get this fixed, I'll look into it in the next couple weeks. Iirc there were lots different setups to test and I'll need to refamiliarize myself with them.

**Comment by parkovski** [comment]
Please don't bot, I'm getting to it.

**Comment by justinmk** [comment]
nuisance bot

File: src/win/tty.c [hunk 4]
```diff
@@ -835,7 +898,8 @@ void uv_process_tty_read_raw_req(uv_loop_t* loop, uv_tty_t* handle,
                                          utf16_buffer,
                                          2,
                                          &handle->tty.rd.last_key[prefix_len],
-                                         sizeof handle->tty.rd.last_key,
+                                         sizeof handle->tty.rd.last_key
+                                           - prefix_len,
```
Comment (parkovski) [inline src/win/tty.c:902]: Added `- prefix_len` here as this looked potentially dangerous.

**Comment by justinmk** [comment]
nuisance bot

**Comment by vtjnash** [comment]
Did you get a chance to resolve https://github.com/libuv/libuv/pull/2535#issuecomment-580394201? (https://github.com/libuv/libuv/pull/2535#discussion_r369175125)

**Comment by bnoordhuis** [comment]
Seeing there's been no follow-up I'll go ahead and close this.

**Pull request closed without merging**